rfd/00XX-tctl-sso-configure-command.md

```diff
@@ -0,0 +1,205 @@
+---
+authors: Krzysztof Skrzętnicki <krzysztof.skrzetnicki@goteleport.com>
+state: draft
+---
+
+# RFD XX - `tctl sso configure` command
+
+## What
+
+This RFD proposes a new subcommand for the `tctl` tool: `sso configure`. This is a convenience command to help with the
+configuration of auth connectors. The command is an alternative to directly editing the auth connector resource files
+and provides automated validation and sanity checks.
+
+Not everything can be checked by the tool, as the overall validity of the SSO setup depends in part on the IdP
+configuration which is invisible for us. This is why the output of the command should be tested with `tctl sso test`
+command, e.g. by piping the output from one to another: `tctl sso configure ... | tctl sso test`.
+
+The input is provided exclusively via command flags. In particular, no "guided" or "interactive" experience is provided.
+This is a deliberate choice, as the web UI admin interface will be a better hosting environment for such a feature.
+
+Ultimately we would like to support all providers for which we have
+detailed [SSO how-to pages](../docs/pages/enterprise/sso). Initial scope covers only SAML.
+
+## Why
+
+We want to simplify the task of configuring a new auth connector. In contrast with free-form text editing of resource,
+this command line tool will only output well-formed, validated auth connector. Whenever possible, the tool will
+automatically fill in the fields, e.g. by fetching the Proxy public address and calculating the respective webapi
+endpoint.
+
+Additionally, the command provides a foundation for creating similar functionality in web-based UI.
+
+## UX
+
+Each connector kind is served by a separate subcommand. Most subcommand flags are not shared.
+
+```bash
+$ tctl sso configure --help
+
+Create auth connector configuration.
+
+...
+
+Commands:
+  sso configure saml   Configure SAML connector, optionally using a preset. Available presets: [okta onelogin ad adfs]
+  sso configure oidc   Configure OIDC auth connector, optionally using a preset.
+  sso configure github Configure GitHub auth connector.
+```
+
+### Flags: SAML
+
+For `SAML` the commonly used flags will be:
+
+```
+-p, --preset                   Preset. One of: [okta onelogin ad adfs]
+-n, --name                     Connector name. Required, unless implied from preset.
+-e, --entity-descriptor        Set the Entity Descriptor. Valid values: file, URL, XML content. Supplies configuration parameters as single XML instead of individual elements.
+-a, --attributes-to-roles      Sets attribute-to-role mapping in the form 'attr_name,attr_value,role1,role2,...'. Repeatable.
```

Comment:
Looking at the examples, I find this syntax a bit hard to parse without going back and referencing the docs. I know it's tricky because the user is having to specify three or more things at once. There might not be a better solution, but I found myself wondering about alternative syntaxes like `-a group[developer]=access`

Reply:
Unfortunately, both `[` and `]` are special characters in many shells which makes them poor choices. But I'm happy to consider alternative syntaxes that do not have this problem.